### ISSUE TITLE
Add `overwrite` argument to create-ome-zarr-structure task

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-tasks-core#485]
An issue that came up multiple times that has been confusing to users (cc @nrepina @al-mcintyre ) was that the Create OME Zarr structure tasks fails if the zarr file already exists. This comes up most often when a user is trying parameters, but something fails and they want to rerun the workflow.

Thus, the `Create OME Zarr structure` task should have an `overwrite` flag that is off by default. If it's set to True, any Zarr file that would match the new plate name is removed. But not other Zarr files in that folder or other files in the output folder.
(Note: Let's not delete the whole output folder. The user may be putting different outputs of different workflows in the same output folder)

If it's set to false and the output file already exist, we should throw a better error than the current zarr error.

(this currently occurs a lot, will occur less once we deploy running subsets of workflows, see https://github.com/fractal-analytics-platform/fractal-server/pull/784)

**Comment by tcompa** [comment]
Default: `overwrite=False`

This is one of the "fresh-copy" kind of overwriting:
* If the zarr group does not exist, proceed as usual (independently on `overwrite`)
* If the zarr group already exists and `overwrite=False`, raise an appropriate error and exit.
* If the zarr group already exists and `overwrite=True`, all its content should be removed before proceeding (ideally with zarr-API calls, rather than actual filesystem removals).

**Comment by tcompa** [comment]
To be checked: what's the behavior of `write_elem` in case the element is already present? Does it expose some kind of `overwrite` option?

See https://anndata.readthedocs.io/en/latest/generated/anndata.experimental.write_elem.html

**Comment by tcompa** [comment]
Documenting the behavior of [`zarr.group`](https://zarr.readthedocs.io/en/stable/api/hierarchy.html#zarr.hierarchy.group):

Script:
```python
import zarr

group_1 = zarr.group("/tmp/group.zarr")
group_1.attrs.put({"key": "value"})
print(f"{group_1.store.path=}")
print(f"{group_1.attrs.asdict()=}")

group_2 = zarr.group("/tmp/group.zarr", overwrite=True)
print(f"{group_2.store.path=}")
print(f"{group_2.attrs.asdict()=}")
```
output
```
group_1.store.path='/tmp/group.zarr'
group_1.attrs.asdict()={'key': 'value'}
group_2.store.path='/tmp/group.zarr'
group_2.attrs.asdict()={}
```

**Comment by tcompa** [comment]
Here is a slightly more complex example (based on the `mode` parameter of [`zarr.open_group`](https://zarr.readthedocs.io/en/stable/api/hierarchy.html#zarr.hierarchy.open_group)), that achieves the expected goal:

```python
import zarr
import os

def zarr_open_group_with_overwrite(*, path, overwrite):
    if overwrite:
        mode = "w"
    else:
        mode = "w-"
    return zarr.open_group(path, mode=mode)


path = "/tmp/group.zarr"

print("# Group does not exist")
assert not os.path.isdir(path)
group_1 = zarr_open_group_with_overwrite(path=path, overwrite=False)
group_1.attrs.put({"key": "value"})
print(f"{group_1.store.path=}")
print(f"{group_1.attrs.asdict()=}")
print()

print("# Group already exists, overwrite=True")
group_2 = zarr_open_group_with_overwrite(path=path, overwrite=True)
print(f"{group_2.store.path=}")
print(f"{group_2.attrs.asdict()=}")
print()

print("# Group already exists, overwrite=False")
try:
    open_group_3 = zarr.open_group("/tmp/open_group.zarr", mode="w-")
except zarr.errors.ContainsGroupError as e:
    print(f"group_3 was not created. Original error:\n\t{str(e)}")
```

With output
```
# Group does not exist
group_1.store.path='/tmp/group.zarr'
group_1.attrs.asdict()={'key': 'value'}

# Group already exists, overwrite=True
group_2.store.path='/tmp/group.zarr'
group_2.attrs.asdict()={}

# Group already exists, overwrite=False
group_3 was not created. Original error:
	path '' contains a group
```

**Comment by tcompa** [comment]
I think the last version to [the last comment](https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/485#issuecomment-1695696556) (warning, I edited it a few times) works as intended.

The `mode` values listed in [`open_group`](https://zarr.readthedocs.io/en/stable/api/hierarchy.html#zarr.hierarchy.open_group) are:
> - ‘r’ means read only (must exist);
> - ‘r+’ means read/write (must exist);
> - ‘a’ means read/write (create if doesn’t exist);
> - ‘w’ means create (overwrite if exists);
> - ‘w-’ means create (fail if exists).

And the relevant ones for us (upon group creation) are:
* `w`, if we want to overwrite an existing group
* `w-` if we do not want to overwrite an existing group

Note that `a` is not relevant upon group creation, since it would clearly not comply with an `overwrite=False` scenario but also it would read-in an existing zarr group when `overwrite=True` (without the "full-refresh" that we need in this case).

The intended behaviors are achieved:
1. If `overwrite=False` and the group does not exist, we simply create it
2. If `overwrite=True` and the group does not exist, we simply create it
3. If `overwrite=True` and the group already exists, we replaced it with a new empty group
4. If `overwrite=False` and the group already exists, we fail with a relevant error

**Comment by jluethi** [comment]
Looks promising! Thanks for documenting this @tcompa

**Comment by jluethi** [comment]
Makes sense to me for how we're using it. Instead of doing this in every task, having that abstraction seems logical.
Also don't have better naming ideas

**Comment by tcompa** [comment]
The only (minor) improvement that comes to mind is that we could wrap `zarr.open` (https://zarr.readthedocs.io/en/stable/api/convenience.html#zarr.convenience.open) instead of `zarr.open_group`, and we'll end up with a function `zarr_open_with_overwrite` which works for both groups and arrays.

**Comment by tcompa** [comment]
> To be checked: what's the behavior of write_elem in case the element is already present? Does it expose some kind of overwrite option?
> See https://anndata.readthedocs.io/en/latest/generated/anndata.experimental.write_elem.html

I could not find a simple way of instructing `write_elem` to behave as expected, therefore I created another wrapper (`write_elem_with_overwrite` - see https://github.com/fractal-analytics-platform/fractal-tasks-core/commit/66341f1d26fdf1d854cd4061174e45a7f77d8498) where I check the existence of the sub-group explicitly.

**Comment by tcompa** [comment]
This issue is in principle solved by #499 (apart from the lack of tests).

**Comment by jluethi** [comment]
Looks great. On the write_elem wrapper: can we take this opportunity to make this a broader wrapper that handles more of the table writing? Write_elem seems like a low level wrapper function. Would be great if tasks can just call a “write_table” or something like that with some parameters. Currently, many task replicate similar habdling of the table metadata, updating the list of tables, checking for existing tables etc.
This would be a good opportunity to generalize this and have a useful wrapper that sets the correct metadata for those tables, but that someone building a new task then doesn’t have to think about
